### PR TITLE
Internal: Rebuilt the navigator using React [ED-5075]

### PR DESCRIPTION
**Pull request by YitzhakAvm**

### CODE REVIEW

File: assets/dev/js/editor/components/documents/commands/internal/attach-preview.js [hunk 1]
```diff
@@ -1,4 +1,12 @@
+import ElementsCollection from 'elementor-elements/collections/elements';
+
 export class AttachPreview extends $e.modules.CommandInternalBase {
+	slicesToReset = [
```
Comment (matipojo) [inline assets/dev/js/editor/components/documents/commands/internal/attach-preview.js:4]: Move to the services themselves. On document change.

File: assets/dev/js/editor/components/selection/manager.js [hunk 2]
```diff
@@ -74,21 +63,32 @@ export default class Manager extends elementorModules.editor.utils.Module {
...
+			);
 
 			container.view.select();
 		}
+
+		this.updateEnvironment();
+
+		if ( section ) {
+			elementor.activateElementSection( section );
+		}
```
Comment (matipojo) [inline assets/dev/js/editor/components/selection/manager.js:91]: Explain

File: assets/dev/js/editor/document/elements/commands-internal/set-settings.js [hunk 3]
```diff
@@ -25,6 +25,13 @@ export class SetSettings extends $e.modules.editor.CommandContainerInternalBase
 				container.render();
 			}
 		} );
+
+		$e.store.dispatch(
```
Comment (matipojo) [inline assets/dev/js/editor/document/elements/commands-internal/set-settings.js:29]: Go over all elements commands and move the Marionette to a separate method:
`updateMarionetteStore`

File: assets/dev/js/editor/document/elements/component.js [hunk 4]
```diff
@@ -15,6 +15,134 @@ export default class Component extends ComponentBase {
...
+
+				// Act
+				for ( const containerId of containerIds ) {
+					// Set settings of the current element
+					if ( state[ containerId ] ) {
+						state[ containerId ].settings = { ...state[ containerId ].settings, ...settings };
+					}
+				}
+			},
+			change: ( state, { payload } ) => {
```
Comment (matipojo) [inline assets/dev/js/editor/document/elements/component.js:89]: Explain the difference.

File: assets/dev/js/editor/document/elements/component.js [hunk 5]
```diff
@@ -15,6 +15,134 @@ export default class Component extends ComponentBase {
 		return this.importCommands( commandsInternal );
 	}
 
+	defaultStates() {
```
Comment (matipojo) [inline assets/dev/js/editor/document/elements/component.js:18]: Create a design doc for the store & it's reducers.

File: assets/dev/js/editor/editor-base.js [hunk 6]
```diff
@@ -561,10 +572,13 @@ export default class EditorBase extends Marionette.Application {
 		this.addRegions( {
 			navigator: {
 				el: '#elementor-navigator',
-				regionClass: Navigator,
+				regionClass: NavigatorRegion,
 			},
 		} );
 
+		this.navigator = $e.components.register( new NavigatorComponent( { manager: this.navigator } ) );
+		this.navigator.elements = $e.components.get( 'navigator/elements' );
```
Comment (matipojo) [inline assets/dev/js/editor/editor-base.js:580]: Define as BC

File: assets/dev/js/editor/elements/views/base.js [hunk 7]
```diff
@@ -119,17 +119,21 @@ BaseElementView = BaseContainer.extend( {
 
 	getContainer() {
 		if ( ! this.container ) {
-			const settingsModel = this.model.get( 'settings' );
+			const elType = this.model.get( 'elType' ),
+				settingsModel = this.model.get( 'settings' );
 
 			this.container = new elementorModules.editor.Container( {
-				type: this.model.get( 'elType' ),
+				type: elType,
```
Comment (matipojo) [inline assets/dev/js/editor/elements/views/base.js:126]: Why?

File: assets/dev/js/editor/elements/views/widget.js [hunk 8]
```diff
@@ -51,6 +51,16 @@ WidgetView = BaseElementView.extend( {
 		return elementor.hooks.applyFilters( 'elements/widget/behaviors', behaviors, this );
 	},
 
+	/**
+	 * Widget usually doesn't contain elements, unless it's in nested state. Therefore we should inform its container
+	 * using this method that it has no children by default.
+	 *
+	 * @returns {boolean}
+	 */
+	hasChildrenByDefault: function() {
```
Comment (matipojo) [inline assets/dev/js/editor/elements/views/widget.js:60]: Move to the new widget model.

**Review by matipojo — CHANGES_REQUESTED** [review]

File: assets/dev/js/editor/document/elements/component.js [hunk 9]
```diff
@@ -15,6 +15,134 @@ export default class Component extends ComponentBase {
...
+		const defaultReducers = {
+			add: ( state, { payload } ) => {
```
Comment (matipojo) [inline assets/dev/js/editor/document/elements/component.js:20]: Move all reducers to thier commands

File: assets/dev/js/editor/document/elements/commands-internal/set-settings.js [hunk 10]
```diff
@@ -33,6 +33,13 @@ export class SetSettings extends $e.modules.editor.CommandContainerInternalBase
 				} )
 			);
 		} );
+
+		$e.store.dispatch(
+			$e.store.get( 'document/elements' ).actions.settings( {
+				containerIds: containers.map( ( container ) => container.id ),
+				settings: { ...settings },
+			} )
+		);
```
Comment (matipojo) [inline assets/dev/js/editor/document/elements/commands-internal/set-settings.js:42]:
```suggestion
```

File: assets/dev/js/editor/elements/views/base.js [hunk 11]
```diff
@@ -841,13 +844,9 @@ BaseElementView = BaseContainer.extend( {
 			return;
 		}
 
-		if ( options.scrollIntoView ) {
```
Comment (matipojo) [inline assets/dev/js/editor/elements/views/base.js:844]: Why?

File: assets/dev/js/editor/elements/views/widget.js [hunk 12]
```diff
@@ -36,6 +36,61 @@ const WidgetView = BaseWidget.extend( {
...
+		return false;
+	},
+
+	getEditButtons: function() {
```
Comment (matipojo) [inline assets/dev/js/editor/elements/views/widget.js]: Check if it exists in the new widget base

File: assets/dev/js/editor/regions/navigator/commands-hooks/ui/editor/documents/attach-preview/navigator-initialize.js [hunk 13]
```diff
@@ -0,0 +1,27 @@
...
+	getId() {
+		return 'navigator-initialize';
+	}
+
+	getConditions() {
+		return elementor.documents.getCurrent().config.panel.has_elements;
+	}
+
+	apply( args ) {
+		const { region } = $e.components.get( 'navigator' );
```
Comment (matipojo) [inline assets/dev/js/editor/regions/navigator/commands-hooks/ui/editor/documents/attach-preview/navigator-initialize.js:17]: Reset folding state.

File: assets/dev/js/editor/regions/navigator/component.js [hunk 14]
```diff
@@ -1,7 +1,28 @@
...
+	__construct( args ) {
+		super.__construct( args );
+
+		this.region = args.manager;
+
+		this.elements = $e.components.register( new ElementsComponent( { manager: this } ) );
+
+		Object.defineProperty( this, 'indicators', {
+			get() {
+				elementorCommon.helpers.softDeprecated( 'elementor.navigator.indicators', '3.6.0', 'elementor.navigator.region.indicators' );
```
Comment (matipojo) [inline assets/dev/js/editor/regions/navigator/component.js:20]: Replace in all files

```suggestion
				elementorCommon.helpers.softDeprecated( 'elementor.navigator.indicators', '3.8.0', 'elementor.navigator.region.indicators' );
```

File: assets/dev/js/editor/regions/navigator/component.js [hunk 15]
```diff
@@ -16,6 +37,35 @@ export default class Component extends ComponentBase {
 		return this.importCommands( commands );
 	}
 
+	defaultCommandsInternal() {
+		return this.importCommands( commandsInternal );
+	}
+
+	defaultStates() {
+		return {
+			folding: {
```
Comment (matipojo) [inline assets/dev/js/editor/regions/navigator/component.js:46]: Move to the commands

**Review by matipojo — CHANGES_REQUESTED** [review]

File: assets/dev/js/editor/regions/navigator/component.js (hunk 14 above)
Comment (rodolphebertozzo) [reply assets/dev/js/editor/regions/navigator/component.js:20]: See: https://github.com/elementor/elementor/pull/17938

File: assets/dev/js/editor/elements/views/base.js (hunk 11 above)
Comment (ItzhakAvm) [reply assets/dev/js/editor/elements/views/base.js:844]: Check out [this](https://github.com/elementor/elementor/pull/17937/files/24c6737e3856d4e105ba901845e44907832b1e52#diff-1112993f94abb738bc19f7ea5baaa313c82feefda8405861d58a423502cb5d6dR10) change.

File: assets/dev/js/editor/regions/navigator/commands-internal/set-size.js [hunk 16]
```diff
@@ -0,0 +1,27 @@
...
+	 * @property args.size {String|null} size A specific new size.
+	 */
+	apply( { size = null } ) {
+		const { storage } = this.component.region;
+
+		if ( size ) {
+			storage.size.width = size;
+		} else {
+			storage.size.width = storage.size.width || elementorCommon.elements.$body.css( '--e-editor-navigator-width' );
+		}
```
Comment (StyleShit) [inline assets/dev/js/editor/regions/navigator/commands-internal/set-size.js:17]:
```suggestion
		storage.size.width = size ?? storage.size.width ?? elementorCommon.elements.$body.css( '--e-editor-navigator-width' );
```

**Review by StyleShit — CHANGES_REQUESTED** [review]

File: assets/dev/js/editor/views/preview.js [hunk 17]
```diff
@@ -6,6 +6,20 @@ const Preview = BaseSectionsContainerView.extend( {
 	initialize: function() {
 		this.$childViewContainer = jQuery( '<div>', { class: 'elementor-section-wrap' } );
 
+		// Subscribe to the selection state kept in redux.
+		$e.store.selector(
```
Comment (StyleShit) [inline assets/dev/js/editor/views/preview.js:10]: createSelector

File: modules/web-cli/assets/js/core/store.js [hunk 18]
```diff
@@ -31,6 +31,19 @@ export default class Store {
 			// Use an empty function instead of empty object since an empty object
 			// isn't a valid reducer value.
 			reducer: () => {},
+			middleware: ( getDefaultMiddleware ) => getDefaultMiddleware( {
```
Comment (StyleShit) [inline modules/web-cli/assets/js/core/store.js:34]: revert everything in this file

File: tests/jest/unit/assets/dev/js/editor/regions/navigator/components/item-title.test.js [hunk 19]
```diff
@@ -0,0 +1,88 @@
+import { render, fireEvent, waitFor } from '@testing-library/react';
```
Comment (StyleShit) [inline tests/jest/unit/assets/dev/js/editor/regions/navigator/components/item-title.test.js:1]:
```suggestion
import { render, fireEvent } from '@testing-library/react';
```

**Review by StyleShit — CHANGES_REQUESTED** [review]

File: .grunt-config/webpack.alias.js [hunk 20]
```diff
@@ -26,6 +26,7 @@ module.exports = {
 			'elementor-utils': path.resolve( __dirname, '../assets/dev/js/utils' ),
 			'elementor-validator': path.resolve( __dirname, '../assets/dev/js/editor/components/validator' ),
 			'elementor-views': path.resolve( __dirname, '../assets/dev/js/editor/views' ),
+			'elementor-tests-qunit': path.resolve( __dirname, '../tests/qunit' ),
```
[the remark this answers is not in the document]
Comment (StyleShit) [reply .grunt-config/webpack.alias.js:29]: remove

**Pull request closed without merging**